### PR TITLE
add support of user-defined endpoints to SwitchLBRule

### PR DESCRIPTION
**Pull request by qiutingjun**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

Add support of user-defined endpoints so that  SwitchLBRule could be used to vms with static ip created by kubevirt.

### Which issue(s) this PR fixes:
Fixes #(issue-number)

### <samp>🤖 Generated by Copilot at 36774d0</samp>

This pull request enhances the switch load balancer rule feature by improving the code quality, adding a new field to specify endpoints, and adding a new function to generate endpoints from services. It modifies `pkg/controller/switch_lb_rule.go` and `pkg/apis/kubeovn/v1/types.go`.

### <samp>🤖 Generated by Copilot at 36774d0</samp>

> _`SwitchLBRuleSpec`_
> _Has new field `Endpoints`_
> _Optional, not fixed_

### <samp>🤖 Generated by Copilot at 36774d0</samp>

*  Add a new field `Endpoints` to the type `SwitchLBRuleSpec` to allow manual specification of endpoints for switch load balancer rules ([link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aR969))
* Rename the function `genSvcName` to `generateSvcName` for consistency and clarity ([link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L28-R28), [link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L189-R217))
* Refactor the function `handleAddOrUpdateSwitchLBRule` to create or update the service and endpoints for the switch load balancer rule based on the `Endpoints` field ([link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L130-R140), [link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L138-R182), [link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L159-R188), [link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L175-R210))
* Refactor the function `generateHeadlessService` to simplify variable declarations and return the service object directly ([link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L217-R256), [link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L233), [link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L242-L243), [link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L252-R285))
* Add a new function `generateEndpoints` to create or update the endpoints object for the switch load balancer rule with the updated addresses and ports ([link](https://github.com/kubeovn/kube-ovn/pull/2777/files?diff=unified&w=0#diff-1bbd8fc68082905bd8c8204a29a746985193081fb932bedaa86f4f25abc2dc02L267-R358))

### CODE REVIEW

File: pkg/controller/switch_lb_rule.go [hunk 1]
```diff
@@ -127,36 +127,65 @@ func (c *Controller) runAddSwitchLBRuleWorker() {
...
 		}
 		return err
 	}
 
-	needToCreate := false
-	name := genSvcName(slr.Name)
-	oldSvc, err := c.config.KubeClient.CoreV1().Services(slr.Spec.Namespace).Get(context.Background(), name, metav1.GetOptions{})
-	if err != nil {
+	svcName = generateSvcName(slr.Name)
+	if oldSvc, err = c.config.KubeClient.CoreV1().Services(slr.Spec.Namespace).Get(context.Background(), svcName, metav1.GetOptions{}); err != nil {
```
Comment (zhangzujian) [inline pkg/controller/switch_lb_rule.go:148]:
```suggestion
	if oldSvc, err = c.servicesLister.Services(slr.Spec.Namespace).Get(svcName); err != nil {
```

File: pkg/controller/switch_lb_rule.go [hunk 2]
```diff
@@ -127,36 +127,65 @@ func (c *Controller) runAddSwitchLBRuleWorker() {
...
+				err = fmt.Errorf("failed to create endpoints '%s', err: %v", eps, err)
+				klog.Error(err)
+				return err
+			}
+		} else {
+			if _, err = c.config.KubeClient.CoreV1().Endpoints(slr.Spec.Namespace).Update(context.Background(), eps, metav1.UpdateOptions{}); err != nil {
+				err = fmt.Errorf("failed to update endpoints '%s', err: %v", eps, err)
+				klog.Error(err)
+				return err
+			}
```
Comment (zhangzujian) [inline pkg/controller/switch_lb_rule.go:175]: Ensure the ep exists before the update.

**Comment by zhangzujian** [comment]
Switch LB rule with custom endpoints does not work as expected, please fix it.

**Comment by qiutingjun** [comment]
> Switch LB rule with custom endpoints does not work as expected, please fix it.

could u show me the case which does not work as expected？i've tested and headless svc && ep  be maintained as expect.

**Comment by qiutingjun** [comment]
> > could u show me the case which does not work as expected？i've tested and headless svc && ep be maintained as expect.
> 
> Create a switch lb rule with custom endpoints, the k8s endpoints created by kube-ovn-controller has no address sets.

did u update the crd of switch-lb-rules.kubeovn.io?

**Comment by qiutingjun** [comment]
> > could u show me the case which does not work as expected？i've tested and headless svc && ep be maintained as expect.
> 
> Create a switch lb rule with custom endpoints, the k8s endpoints created by kube-ovn-controller has no address sets.

ok,conflict of selector & endpoints caused this problem.i'll discard the selector if endpoints is set.

File: pkg/controller/switch_lb_rule.go [hunk 3]
```diff
@@ -127,36 +127,95 @@ func (c *Controller) runAddSwitchLBRuleWorker() {
...
-			err := fmt.Errorf("failed to create service '%s', err: %v", svc, err)
+			needToRevertEps = true
+			err = fmt.Errorf("failed to create service '%s', err: %v", svc, err)
 			klog.Error(err)
 			return err
 		}
 	} else {
 		if _, err = c.config.KubeClient.CoreV1().Services(slr.Spec.Namespace).Update(context.Background(), svc, metav1.UpdateOptions{}); err != nil {
-			err := fmt.Errorf("failed to update service '%s', err: %v", svc, err)
+			needToRevertEps = true
```
Comment (bobz965) [inline pkg/controller/switch_lb_rule.go:217]: 如果确定是要更新成功，应该一直重试更新，而非回滚

File: pkg/controller/switch_lb_rule.go [hunk 4]
```diff
@@ -127,36 +127,95 @@ func (c *Controller) runAddSwitchLBRuleWorker() {
...
+				return err
+			}
+		} else {
+			if _, err = c.config.KubeClient.CoreV1().Endpoints(slr.Spec.Namespace).Update(context.Background(), eps, metav1.UpdateOptions{}); err != nil {
+				err = fmt.Errorf("failed to update endpoints '%s', err: %v", eps, err)
+				klog.Error(err)
+				return err
+			}
+		}
+		slr.Spec.Selector = nil
```
Comment (bobz965) [inline pkg/controller/switch_lb_rule.go:188]: 添加一些注释用于标识spec的Endpoints 和 Selector 在业务逻辑上是是二选一生效的，Endpoints 用于多网卡pod的svc selector机制作用不到的情况，该情况下endpoint只能基于该维护

**Review by bobz965 — CHANGES_REQUESTED** [review]

File: pkg/controller/switch_lb_rule.go (hunk 3 above)
Comment (bobz965) [reply pkg/controller/switch_lb_rule.go:217]: 除非有些操作导致无法重入，必须清理，否则直接重试应该就可以

**Comment by bobz965** [comment]
需要更新下文档 区分slr 的两种用法，和这个注释基本一致 `添加一些注释用于标识spec的Endpoints 和 Selector 在业务逻辑上是是二选一生效的，Endpoints 用于多网卡pod的svc selector机制作用不到的情况，该情况下endpoint只能基于该维护
`

**Review by bobz965 — APPROVED** [review]

**Comment by patriziobassi** [comment]
Hi, any documentation about how to use this feature, such as a doc page or yaml example? Thank you

**Pull request merged**